### ISSUE TITLE
Allow integration authors to deprecate integrations and data streams

### ISSUE THREAD

**Issue by mtojek** [elastic/package-spec#227]
## Summary

Integration authors want to deprecate packages or data streams as they are no longer relevant, or when they are superseded by new options.

Fleet and integrations should support deprecated packages and data streams, and provide a clear and consistent UX around deprecated configurations in integrations that steers users to migrate to new and improved functionality.  

## Implementation plan

See sub-issues for implementation plan

---

<details>
  <summary>Original description</summary>
  With 8.x.x we may want to deprecate some package or data streams and create new alternatives. We need to discuss many things around, including publishing and installation rules of deprecated policies and "deprecated" captions.  
content
</details>

**Comment by ruflin** [comment]
One core feature related to this needed in Fleet is that even if a package is removed from the registry, Kibana will still keep working assuming the package was installed before. The package cannot be upgraded anymore or reinstalled, but things keep working. Ideally Fleet would know from which registry it was installed and can indicate, that the package does not exists anymore in the remote registry.

**Comment by ruflin** [comment]
As a follow up, we likely have to expand the deprecation discussion to more granular parts which could be deprecated:

* integration policy
* data stream
* field
* ?

Each likely needs its own story on how it can be deprecated. Ideally we provide framework and guideline for each. Especially around fields discussions are already happening in various places.

**Comment by mtojek** [comment]
As an integration policy, do you mean a policy template?

In terms of spec changes, it should be relatively easy.

Changes in data streams and packages can be quick wins, as they both contain manifests. We can augment these files with a special "deprecation" structure.
Theoretically, we can introduce a similar flag to "field" (like we do in Beats).

I don't have any preference regarding Kibana/Fleet behavior. I guess we could redirect users of one package to another? or just suggest it to them?

**Comment by ruflin** [comment]
@mtojek Yes, policy templates. Should we discuss each of these topics in a separate issue as some might need work in Fleet, others don't and we can close one by one.

**Comment by jsoriano** [comment]
> @mtojek Yes, policy templates. Should we discuss each of these topics in a separate issue as some might need work in Fleet, others don't and we can close one by one.

 @ruflin what would be the use case of deprecating policy templates without deprecating the data_stream that contains it?

**Comment by klacabane** [comment]
Hi there, our team may have a use case for deprecated inputs.
 
The packages for the stack components (elasticsearch/kibana/logstash) currently have two inputs, one for metrics collection via a metricbeat module and one for logs. The metricbeat collection will eventually be superseded by a new approach that uses different metrics schema, assets and collection method, while the logs collection remains the same. When the new data streams are ready the package would allow a transition period where both metrics collection are available, with the metricbeat one marked as deprecated.
Once the deprecation period is done and we ship a new version that removes the input, it'd be nice to surface that in the UI when trying to upgrade to that version (maybe with a delta of the current/next version inputs, or an additional metadata flag) so that users can take actions before committing to a breaking change

**Comment by jsoriano** [comment]
This sounds like a valid use case for this, yes :+1:

I guess that in this case it would be nice to mark the input as deprecated as soon as the other one is ready, with a `replacedBy` rule like the ones proposed in https://github.com/elastic/package-spec/issues/227#issuecomment-946495979. Then users would have a period while they can migrate.
I think the UI should warn about this as soon as it sees the deprecation notice.

**Comment by klacabane** [comment]
It's possible that the replacement won't be an elastic-agent input that collects the metrics, but a push-based model where the stack components would send metrics to apm server. The package would only be a shell for the assets in that case. The `replacedBy` setting could potentially point to the apm-server package but that doesn't sound like the right way.
We probably don't need anything sophisticated for the deprecation metadata here besides a way to provide guidance on the next steps (a message like "the metrics collection now require a running apm-server and updates to the elasticsearch.yml file. Refer to the package documentation for the steps to follow").

What we could use is a dependency relationship with another package. If the new metrics collection needs another package to be setup, do we currently have a way to surface that ?

**Comment by jsoriano** [comment]
The need for package deprecation appeared again in https://github.com/elastic/integrations/issues/3262#issuecomment-1569806935.

**Comment by jsoriano** [comment]
> I still need to write a ticket to cover the docs about the new deprecation lifecycle. [@jsoriano](https://github.com/jsoriano) what's the correct repo for these docs?

I would add something for developers to https://github.com/elastic/elastic-package/tree/main/docs/howto, and open an issue in https://github.com/elastic/ingest-docs for user facing docs.

**Comment by criamico** [comment]
@jlind23 the work is on track, only https://github.com/elastic/package-spec/issues/956 is taking longer as the integration team is currently helping me with real case scenarios and testing. That one could take longer to be finished, but the all the other issue are close to be completed.

**Comment by jlind23** [comment]
@criamico Should we pull someone from the docs team to start looking into https://github.com/elastic/ingest-docs/issues/1879?
